nomos-services/data-availability/sampling/src/network/adapters/libp2p.rs

```diff
@@ -0,0 +1,127 @@
+// std
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::pin::Pin;
+
+use rand::prelude::*;
+use rand_chacha::ChaCha20Rng;
+
+// crates
+use futures::{Stream, StreamExt};
+use libp2p_identity::PeerId;
+use tokio::sync::oneshot;
+// internal
+use crate::network::NetworkAdapter;
+use nomos_core::da::BlobId;
+use nomos_da_network_core::SubnetworkId;
+use nomos_da_network_service::backends::libp2p::validator::{
+    DaNetworkEvent, DaNetworkEventKind, DaNetworkMessage, DaNetworkValidatorBackend, SamplingEvent,
+};
+use nomos_da_network_service::{DaNetworkMsg, NetworkService};
+use overwatch_rs::services::relay::OutboundRelay;
+use overwatch_rs::services::ServiceData;
+use overwatch_rs::DynError;
+use subnetworks_assignations::MembershipHandler;
+
+#[derive(Debug, Clone)]
+pub struct DaNetworkSamplingSettings {
+    pub num_samples: u16,
+    pub subnet_size: SubnetworkId,
+}
+
+pub struct SamplingContext {
+    blob_id: BlobId,
+    subnets: Vec<SubnetworkId>,
+}
+
+pub struct Libp2pAdapter<Membership>
+where
+    Membership: MembershipHandler<NetworkId = SubnetworkId, Id = PeerId>
+        + Debug
+        + Clone
+        + Send
+        + Sync
+        + 'static,
+{
+    settings: DaNetworkSamplingSettings,
+    network_relay: OutboundRelay<
+        <NetworkService<DaNetworkValidatorBackend<Membership>> as ServiceData>::Message,
+    >,
+    pending_sampling: HashMap<BlobId, SamplingContext>,
+    // TODO: is there a better place for this? Do we need to have this even globally?
+    // Do we already have some source of randomness already?
+    rng: ChaCha20Rng,
+}
+
+#[async_trait::async_trait]
+impl<Membership> NetworkAdapter for Libp2pAdapter<Membership>
+where
+    Membership: MembershipHandler<NetworkId = SubnetworkId, Id = PeerId>
+        + Debug
+        + Clone
+        + Send
+        + Sync
+        + 'static,
+{
+    type Backend = DaNetworkValidatorBackend<Membership>;
+    type Settings = DaNetworkSamplingSettings;
+
+    async fn new(
+        settings: Self::Settings,
+        network_relay: OutboundRelay<<NetworkService<Self::Backend> as ServiceData>::Message>,
+    ) -> Self {
+        Self {
+            settings,
+            network_relay,
+            pending_sampling: HashMap::new(),
+            rng: ChaCha20Rng::from_entropy(),
+        }
+    }
+
+    async fn start_sampling(&mut self, blob_id: BlobId) -> Result<(), DynError> {
+        let mut ctx: SamplingContext = SamplingContext {
+            blob_id: (blob_id),
+            subnets: vec![],
+        };
+        for _i in 1..self.settings.num_samples {
```

Comment:
why from 1?, this would result in off by 1 error I think.

Reply:
Actually, I just realize, that this is wrong. Iterating like this may produce the same number twice (or more). Which is undesirable. 
I think instead of itereting, you need to generate the subnetwork ids using somehting like [`random::sample`](https://docs.rs/rand/latest/rand/distributions/trait.Distribution.html#method.sample_iter). So `N` different subnetwork are selected, and with that you can build the message.

Reply:
You're right. Moved to use the suggestion you gave.